crates/pallet-domains/src/staking_epoch.rs

```diff
@@ -547,6 +560,108 @@ fn finalize_nominator_deposit<T: Config>(
     Ok(())
 }
 
+pub(crate) fn do_finalize_slashed_operators<T: Config>(
+    domain_id: DomainId,
+) -> Result<(), TransitionError> {
+    let mut total_slashed_amount = BalanceOf::<T>::zero();
+    for (operator_id, slash_info) in PendingSlashes::<T>::take(domain_id).unwrap_or_default() {
+        let slashed_amount = slash_operator::<T>(operator_id, slash_info)?;
+        total_slashed_amount = total_slashed_amount
+            .checked_add(&slashed_amount)
+            .ok_or(TransitionError::BalanceOverflow)?;
+    }
+
+    // TODO: mint the total slashed amount to treasury
+
+    Ok(())
+}
+
+fn thaw_and_burn_frozen_balance<T: Config>(
+    id: &FungibleFreezeId<T>,
+    from: &T::AccountId,
+) -> Result<(), TransitionError> {
+    // thaw the frozen amount and burn
+    let locked_amount = T::Currency::balance_frozen(id, from);
+    T::Currency::thaw(id, from).map_err(|_| TransitionError::RemoveLock)?;
+    T::Currency::burn_from(from, locked_amount, Precision::Exact, Fortitude::Force)
+        .map_err(|_| TransitionError::BalanceBurn)?;
+
+    Ok(())
+}
+
+#[derive(TypeInfo, Debug, Encode, Decode, Clone, PartialEq, Eq)]
+pub struct PendingOperatorSlashInfo<NominatorId, Balance> {
+    pub unlocking_nominators: Vec<PendingNominatorUnlock<NominatorId, Balance>>,
+}
+
+fn slash_operator<T: Config>(
+    operator_id: OperatorId,
+    slash_info: PendingOperatorSlashInfo<NominatorId<T>, BalanceOf<T>>,
+) -> Result<BalanceOf<T>, TransitionError> {
+    Operators::<T>::try_mutate_exists(operator_id, |maybe_operator| {
+        // take the operator so this operator info is removed once we slash the operator.
+        let operator = maybe_operator
+            .take()
+            .ok_or(TransitionError::UnknownOperator)?;
+
+        let freeze_identifier = T::FreezeIdentifier::staking_freeze_id(operator_id);
+
+        let total_stake = operator
+            .current_total_stake
+            .checked_add(&operator.current_epoch_rewards)
+            .ok_or(TransitionError::BalanceOverflow)?;
+
+        let unlocking_withdrawals = slash_info
+            .unlocking_nominators
+            .into_iter()
+            .map(|pending_unlock| pending_unlock.balance)
+            .try_fold(BalanceOf::<T>::zero(), |acc, pending_withdrawal| {
+                acc.checked_add(&pending_withdrawal)
+                    .ok_or(TransitionError::BalanceOverflow)
+            })?;
+
+        let mut pending_deposits = BalanceOf::<T>::zero();
+
+        Nominators::<T>::drain_prefix(operator_id).try_for_each(|(nominator_id, _)| {
+            thaw_and_burn_frozen_balance::<T>(&freeze_identifier, &nominator_id)?;
+
+            // if there was a pending deposit from this nominator,
+            // then account that deposit in the slashed amount since this new deposit
+            // was burned above
+            if let Some(deposit) = PendingDeposits::<T>::take(operator_id, nominator_id) {
```

Comment:
This seems unfair, the nominator's stake has not participated in the staking yet but got slashed due to the misbehavior that was conducted by the operator before.